### ISSUE TITLE
Race condition in maps load?

### ISSUE THREAD

**Issue by awong-dev** [findthemasks/findthemasks#287]
I'm getting this periodically if I hard refresh the page during local testing.  I think there's a race between the script blob in index.html that loads the maps API and our JS file.

```
locations-list-map.js:382 Uncaught ReferenceError: google is not defined
    at initMap (locations-list-map.js:382)
    at Object.success (locations-list-map.js:288)
    at c (jquery.min.js:2)
    at Object.fireWith [as resolveWith] (jquery.min.js:2)
    at l (jquery.min.js:2)
    at XMLHttpRequest.<anonymous> (jquery.min.js:2)
```

**Comment by awong-dev** [comment]
FYI, this seems to only happen if the JS loads at different rates. Mashing reload on prod doesn't reproduce.  Using local.findthemasks.com:8080 with a local server bound to 8080 does seem to give enough gap that the race occurs about 1/3 of the time on a full reload for me.

The code needs to not call initMap() until google maps api has initialized. Not sure the right way to do that.

We've had this bug live for a while so it's not critical, but it's real and likely explains reports of folks on mobile getting blank donation lists as mobile has a less uniform network layer.

**Comment by bamnet** [comment]
It looks like the Google Maps API script tag is potentially loading late (`defer` tag?) so there's no guarantee when it will be ready.  Suggest making it load sync before the main JS blob executes (easy fix) or use a callback to trigger all of the features that require google.maps (cleaner fix).

**Comment by drwestco** [comment]
Maps script is loaded with defer. locations-list-map is loaded as a module (implied defer). Deferred scripts are supposed to execute in the order they appear in the doc, though maybe explicit defer and modules are queued separately?

Another option would be to move the script tag creation for maps from the html file into locations-list-map and use the callback param to trigger map initialization. Could then avoid downloading the script at all for no-map scenarios.

**Comment by patricknelson** [comment]
For what it's worth, you can define a `&callback=[function name]` in the URL of the Google Maps API JS if you wanted to ensure that dependencies wait for maps to be fully ready prior to calling.

I tried doing this, but that's why I was so bent on using module-level vars to maintain state since then at least we could a little more easily invoke a single function that would then kick everything else off (unfortunately due to the inter-dependencies, it was a little tough to workaround). E.g. Things that don't necessarily _seem_ at first glance to require maps (e.g. the `createContent()` function) actually did depend on maps, so basically entire page load had to wait for maps first, thus you encounter this error.

Originally I ended up defining a function like `triggerMapsApiLoaded` which _then_ would kick off `initMaps(...)` with it's necessary parameters and etc. I was able to queue those up in the context of the jQuery DOM loaded function by using a special queuing function like

```js
let mapQueueCallbacks = [];
function onMapsLoaded(callback) {
  if (typeof google !== 'undefined') {  
    // fire immediately
    callback();
  } else {
    mapQueueCallbacks.push(callback);
  }
}

function triggerMapsApiLoaded() {
  for(let i = 0; i < mapQueueCallbacks.length; i++) {
    let callback = mapQueueCallbacks[i];
    callback();
  }
}
```

Or something along those lines. That way you can always do stuff like this when you know the nested code requires maps or requires something that requires maps, or requires something that was built by something that... requires maps :neckbeard: :

```js

showList && createContent(window.data_by_location);
onMapsLoaded(() => {
  showList && createContent(window.data_by_location);
});

// ... and lower down....

if (showMap) {
  onMapsLoaded(() => {
    initMap(stateFilter);
  })
}
```

... it's hacky AF, but what so far hasn't been? 😏

**Comment by awong-dev** [comment]
Someone wanna throw up a CL?  I like @drwestco 's solution of now downloading themaps api unless needed.Then the center of control still has a single entrt-point in the JS.

**Comment by awong-dev** [comment]
fixed in #296
